### ISSUE TITLE
Issue with execve without fork

### ISSUE THREAD

**Issue by cosinusoidally** [ironmeld/builder-hex0#7]
I ran into this issue when testing out my alternative bootstrap path under builder-hex0 https://github.com/fosslinux/live-bootstrap/pull/407#issuecomment-1890777336 . My cdrun utility seems to cause crashes under builder-hex0. With my tcc_bootstrap_alt path I do not rely on the availability of the cd command, instead I bundle my own utility that will chdir and then execve a command https://github.com/cosinusoidally/tcc_bootstrap_alt/blob/0.2/tcc_js/cdrun.c . I bundle this command so my alt tcc bootstrap can be entirely self contained and not have to rely on any external commands.

As far as I can tell things seem to start going wrong here https://github.com/ironmeld/builder-hex0/blob/8621e56191c635c04fa68687e411016c2b1deab7/builder-hex0-x86-stage2.hex0#L1582
```
#not_forked
8B 60 0C              # mov esp, [eax+0xC]       ; no fork so reset initial stack to same as current process
A1 DE 85 00 00        # mov eax, [next_process_num]
48                    # dec eax
EB 08                 # jmp prepare_stack
```
If I replace `mov esp, [eax+0xC]` with nops that seems to resolve the issue. If applied to this branch of live-bootstrap https://github.com/cosinusoidally/live-bootstrap/tree/builder-hex0-issue  it then seems to run to completion (or at least I assume it will, it got as far as booting Linux before I terminated it). I'm not sure if that is a valid fix though as I'm not that familiar with the builder-hex0 code.

Not yet sure if my hack will break the mes build path (I'm just running that now, but it will take a while to run).

Note my builder-hex0-issue branch has been updated and is now pulling in an updated release of tcc_bootstrap_alt (https://github.com/cosinusoidally/tcc_bootstrap_alt/tree/0.2 which fixes a misalignment issue with malloc)

**Comment by rick-masters** [comment]
Thank you for the bug report. I've been sick for the last few days so I haven't been able to get to it yet. I expect to start working on this on Monday. I'm confident I'll be able to resolve it properly but it's a difficult issue so I don't expect it will be happen right away.

**Comment by rick-masters** [comment]
I'm now sure that I need to remove the line of code you identified. I can't say for sure why it was there but I think I conflated restoring the stack in execve with what exit needs to do.

The field is set by fork but is otherwise not initialized for a new process. It seems to have a leftover valid value in many cases that kept execve from crashing more often. stage0-posix does an exec which has been working ok. But when I initialized the field properly I had to remove the line that sets esp or exec stopped working entirely.
